### PR TITLE
kirkstone: backport chromium from scarthgap

### PR DESCRIPTION
**Pull request by adalessandro**
```
    Merge branch 'scarthgap' into kirkstone
    
    * Backports chromium 131.0.6778.139 to kirkstone.
    
    This relies on the recently added layer branches to satisfy the required
    dependencies:
    
    * Rust: meta-lts-mixins' kirkstone/rust-1.75 branch.
    * Clang: meta-clang's `kirkstone-clang18` branch.
```

* rust
  * https://git.yoctoproject.org/meta-lts-mixins/tree/?h=kirkstone/rust-1.75
  * https://lists.yoctoproject.org/g/yocto-patches/topic/meta_lts_mixins_kirkstone_rust_1_75/109794500

* clang
  * https://github.com/kraj/meta-clang/tree/kirkstone-clang18
  * https://github.com/kraj/meta-clang/pull/1040

Links:
* https://github.com/OSSystems/meta-browser/issues/842
* https://github.com/kraj/meta-clang/issues/1023

### CODE REVIEW

**Comment by adalessandro** [comment]
@MaxIhlenfeldt:

* This is building `chromium` using:
  * https://github.com/kraj/meta-clang/pull/1024
  * We need rust version `>1.74.0` due to `Struct std::ffi::OsStr` which is provided by `meta-lts-mixins` in branch `kirkstone/rust`. In this case, I've tested using rust version `1.77.2` from commit https://git.yoctoproject.org/meta-lts-mixins/commit/?h=kirkstone/rust&id=b1cfa357ba4f046c9f739ceb357fe8a7971cc0d3 due to a few build errors when I used that branch's `HEAD` (debugging pending).

File: meta-chromium/recipes-browser/chromium/rust_%.bbappend
Comment (MaxIhlenfeldt) [inline meta-chromium/recipes-browser/chromium/rust_%.bbappend:1]: Why is this needed? Same question for the changes to `chromium-gn.inc` and `0009-Adjust-the-Rust-build-to-our-needs.patch`. I'd like to avoid reintroducing this `bbappend`.

**Review by MaxIhlenfeldt — COMMENTED** [review]
That was fast, awesome!

> What's the right way to get this merged?

I'd say the first step is merging the `meta-clang` changes, before that we can't merge this PR.

> Note that this PR targets `kirkstone`, but it's based on `master`.

We only create separate branches for Yocto releases that we don't support any more to capture the last supported version. As I understand it, the goal here is to reintroduce support for kirkstone, so the PR should target `master`.

> The README needs to be updated as well. Could you help there perhaps?

Sure! But let's wait until the PR is ready to be merged before updating the README in case something changes during review.

> Is there a testing/CI to be used here?

It'd be great if you could state which build configurations you've tested yourself. @SoumyaWind usually helps with testing for the version updates. But I'd wait with that until the `meta-clang` changes have been merged.

File: meta-chromium/recipes-browser/chromium/rust_%.bbappend
Comment (MaxIhlenfeldt) [reply meta-chromium/recipes-browser/chromium/rust_%.bbappend:1]: Nvm, it seems I looked at an older revision.

**Comment by SoumyaWind** [comment]
I will start testing these changes locally. Will update results once done.

**Comment by SoumyaWind** [comment]
I have completed testing for the following configurations for Poky kirkstone with the changes from this PR and meta-lts-mixins (checked out to Rust version 1.77.2) along with meta-clang from [https://github.com/kraj/meta-clang/pull/1024](https://github.com/kraj/meta-clang/pull/1024)]. Here are the results for reference:

1. **chromium-x11 on qemux86-64 target**: Compiled successfully without any build issues, verified chromium version using the qemux86-64 emulator.
2. **chromium-x11 on qemuarm target**: Compiled successfully without any build issues, verified chromium version using the qemuarm emulator.
3. **chromium-x11 on qemuarm64 target**: Compiled successfully without any build issues, verified chromium version using the qemuarm64 emulator.
4. **chromium-x11 on raspberrypi4-64 target**: Compiled successfully without any build issues, verified chromium version from command line on the Raspberry Pi 4.
5. **chromium-ozone-wayland on qemux86-64 target**: Compiled successfully without any build issues, verified chromium version using the qemux86-64 emulator.

**Comment by adalessandro** [comment]
@SoumyaWind thanks a lot for the testing scenarios covered!
@MaxIhlenfeldt , as per comment https://github.com/kraj/meta-clang/pull/1024#issuecomment-2451851507 the clang major version bump won't be accepted and could be submitted to https://git.yoctoproject.org/meta-lts-mixins/ . Shall we try to push forward in that direction?

**Comment by MaxIhlenfeldt** [comment]
Yes, iirc the initial feedback on the mailing list was positive, so let's do that!

**Comment by adalessandro** [comment]
Backported from scarthgap. The following dependencies are now satisfied from the recently added branches:

* rust
  * https://git.yoctoproject.org/meta-lts-mixins/tree/?h=kirkstone/rust-1.75
  * https://lists.yoctoproject.org/g/yocto-patches/topic/meta_lts_mixins_kirkstone_rust_1_75/109794500

* clang
  * https://github.com/kraj/meta-clang/tree/kirkstone-clang18
  * https://github.com/kraj/meta-clang/pull/1040

**Comment by adalessandro** [comment]
> Yes, iirc the initial feedback on the mailing list was positive, so let's do that!

@MaxIhlenfeldt please review this one now that we have the required dependencies layers available.

File: meta-chromium/recipes-browser/chromium/files/0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch
Comment (MaxIhlenfeldt) [inline meta-chromium/recipes-browser/chromium/files/0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch:1]: I think this shouldn't be necessary when using the `kirkstone/rust` branch of meta-lts-mixins (as opposed to `kirkstone/rust-1.75`), as Rust supports the flag since 1.78. I'll give this PR a spin without this patch.

Comment (adalessandro) [reply meta-chromium/recipes-browser/chromium/files/0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch:1]: I haven't tested using `kirkstone/rust`, however I'd go with a specific version branch, i.e. `kirkstone/rust-1.75` in this case. Branch `kirkstone/rust` is a much more moving target, I wonder if that could bring issues. @MaxIhlenfeldt but you might have a better idea on what's the best approach here :-)

Comment (MaxIhlenfeldt) [reply meta-chromium/recipes-browser/chromium/files/0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch:1]: >  Branch kirkstone/rust is a much more moving target, I wonder if that could bring issues. 

I'd argue this is a good thing, then we don't need to keep updating our docs about which branch to use every time upstream bumps the minimum Rust version.

File: meta-chromium/README.md [hunk 1]
```diff
@@ -79,13 +89,13 @@ Chromium's release model because it often requires recent versions of certain
...
-Because meta-clang's kirkstone branch currently only provides clang 14, we are
-unable to support Chromium versions higher than 120 due to those versions
-requiring a more recent clang version. Also, kirkstone users **must** ensure the
-following:
-* Chromium needs a somewhat recent version of Rust, and for that
-  meta-lts-mixins' **kirkstone/rust-1.68** or **kirkstone/rust-1.70** branch
-  needs to be used.
+
+* Chromium needs a more recent version of Rust than OE Core provides for
+  kirkstone, which is why we depend on meta-lts-mixins' `kirkstone/rust-1.75`
```
Comment (MaxIhlenfeldt) [inline meta-chromium/README.md:94]: Please update this to say `kirkstone/rust`.

**Review by MaxIhlenfeldt — APPROVED** [review]
My build with meta-lts-mixins @ `kirkstone/rust` and without `0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch` was successful :tada:

Please go ahead and remove that patch, and if @rakuco doesn't have any objections we can merge this!

File: meta-chromium/README.md (hunk 1 above)
Comment (adalessandro) [reply meta-chromium/README.md]: Done

**Comment by adalessandro** [comment]
> My build with meta-lts-mixins @ `kirkstone/rust` and without `0013-build-config-compiler-BUILD.gn-Disable-external-clan.patch` was successful 🎉
> 
> Please go ahead and remove that patch, and if @rakuco doesn't have any objections we can merge this!

Great :-) Looking forward to it. Thanks!

**Comment by adalessandro** [comment]
Summary off the diff delta:

```
$ git diff --stat -p origin/scarthgap adalessandro/wip/adalessandro/kirkstone-backport-scarthgap 
 meta-chromium/README.md       | 15 +++++++--------
 meta-chromium/conf/layer.conf |  4 ++--
 2 files changed, 9 insertions(+), 10 deletions(-)

diff --git a/meta-chromium/README.md b/meta-chromium/README.md
index 41ca133..d8b618f 100644
--- a/meta-chromium/README.md
+++ b/meta-chromium/README.md
@@ -82,20 +82,19 @@ host. clang-native from the meta-clang layer is used to build those binaries.
 Additionally, make sure the machine being used to build Chromium is powerful
 enough: a x86-64 machine with at least 16GB RAM is recommended.
 
-### scarthgap-specific requirements
+### kirkstone-specific requirements
 
-The scarthgap OE/Yocto branch is an LTS release, which is often at odds with
+The kirkstone OE/Yocto branch is an LTS release, which is often at odds with
 Chromium's release model because it often requires recent versions of certain
 recipes to build correctly.
 
 This is particularly a problem for the toolchain (i.e. LLVM/clang and Rust).
-Chromium needs a more recent version of Rust than OE Core provides for
-scarthgap, which is why we depend on meta-lts-mixins' `scarthgap/rust` branch.
 
-**Side note: For now, clang 18 provided by meta-clang is recent enough, but at
-some point during scarthgap's LTS lifetime Chromium won't be compilable with
-that version, and we'll have to create e.g. a scarthgap-clang20 branch for
-meta-clang and use that.**
+* Chromium needs a more recent version of Rust than OE Core provides for
+  kirkstone, which is why we depend on meta-lts-mixins' `kirkstone/rust` branch.
+
+* Chromium needs at least clang 18, but meta-clang's kirkstone branch provides
+  clang 14. Thus, we depend on meta-clang's `kirkstone-clang18` branch.
 
 ## PACKAGECONFIG knobs
 
diff --git a/meta-chromium/conf/layer.conf b/meta-chromium/conf/layer.conf
index 9537220..697eae0 100644
--- a/meta-chromium/conf/layer.conf
+++ b/meta-chromium/conf/layer.conf
@@ -9,6 +9,6 @@ BBFILE_PATTERN_chromium-browser-layer := "^${LAYERDIR}/"
 BBFILE_PRIORITY_chromium-browser-layer = "7"
 
 LAYERVERSION_chromium-browser-layer = "1"
-LAYERSERIES_COMPAT_chromium-browser-layer = "scarthgap styhead walnascar"
+LAYERSERIES_COMPAT_chromium-browser-layer = "kirkstone scarthgap styhead walnascar"
 
-LAYERDEPENDS_chromium-browser-layer = "clang-layer core openembedded-layer scarthgap-rust-mixin"
+LAYERDEPENDS_chromium-browser-layer = "clang-layer core openembedded-layer lts-rust-mixin"
```

**Review by rakuco — APPROVED** [review]
lgtm, thank you!

**Comment by MaxIhlenfeldt** [comment]
Thanks again from my side as well!
